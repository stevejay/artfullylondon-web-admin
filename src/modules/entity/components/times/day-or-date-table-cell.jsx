// @flow

import * as React from "react";
import { TableCell, Text } from "grommet";

type Props = {
  +label: string
};

const DayOrDateTableCell = ({ label }: Props) => (
  <TableCell verticalAlign="top">
    <Text>{label}</Text>
  </TableCell>
);

export default DayOrDateTableCell;
