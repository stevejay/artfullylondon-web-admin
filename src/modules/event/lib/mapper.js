import statusType from '_src/domain/types/status-type'
import { entityMapper } from '_src/modules/entity'
import * as eventLib from '_src/shared/lib/event'

export function getInitialValues (event) {
  const isPaid = eventLib.eventIsPaid(event.costType)
  const bookingIsRequired = eventLib.bookingRequired(event.bookingType)
  const hasOccurrenceRange = eventLib.occurrenceHasDateRange(
    event.occurrenceType
  )

  return {
    id: event.id || null,
    status: event.status || statusType.ACTIVE,
    validStatuses: entityMapper.getValidStatusesInitialValue(event.status),
    eventType: event.eventType || '',
    name: event.name || '',
    occurrenceType: event.occurrenceType || '',
    dateFrom: hasOccurrenceRange ? event.dateFrom : null,
    dateTo: hasOccurrenceRange ? event.dateTo : null,
    soldOut: !!event.soldOut,
    costType: event.costType || '',
    costFrom: isPaid ? event.costFrom.toFixed(2) : '',
    costTo: isPaid ? event.costTo.toFixed(2) : '',
    bookingType: event.bookingType || '',
    bookingOpens: bookingIsRequired ? event.bookingOpens : null,
    summary: event.summary || '',
    rating: event.rating ? event.rating.toString() : '3',
    minAge: event.minAge ? event.minAge.toString() : '',
    maxAge: event.maxAge ? event.maxAge.toString() : '',
    eventSeries: event.eventSeries
      ? {
        entityType: event.eventSeries.entityType,
        eventSeriesType: event.eventSeries.eventSeriesType,
        id: event.eventSeries.id,
        name: event.eventSeries.name
      }
      : {},
    venue: event.venue
      ? {
        entityType: event.venue.entityType,
        status: event.venue.status,
        venueType: event.venue.venueType,
        id: event.venue.id,
        name: event.venue.name,
        address: event.venue.address,
        postcode: event.venue.postcode
      }
      : {},
    venueGuidance: event.venueGuidance || '',
    useVenueOpeningTimes: !!event.useVenueOpeningTimes,
    timesRanges: entityMapper.getTimesRangesInitialValue(event.timesRanges),
    openingTimes: entityMapper.getOpeningTimesInitialValue(event.openingTimes),
    additionalOpeningTimes: entityMapper.getAdditionalOpeningTimesInitialValue(
      event.additionalOpeningTimes
    ),
    specialOpeningTimes: entityMapper.getSpecialOpeningTimesInitialValue(
      event.specialOpeningTimes
    ),
    openingTimesClosures: entityMapper.getOpeningTimesClosuresInitialValue(
      event.openingTimesClosures
    ),
    performances: entityMapper.getPerformancesInitialValue(event.performances),
    additionalPerformances: entityMapper.getAdditionalPerformancesInitialValue(
      event.additionalPerformances
    ),
    specialPerformances: entityMapper.getSpecialPerformancesInitialValue(
      event.specialPerformances
    ),
    performancesClosures: entityMapper.getPerformancesClosuresInitialValue(
      event.performancesClosures
    ),
    duration: event.duration || '',
    talents: _getTalentsInitialValue(event.talents),
    audienceTags: event.audienceTags || [],
    mediumTags: event.mediumTags || [],
    styleTags: event.styleTags || [],
    geoTags: event.geoTags || [],
    reviews: event.reviews || [],
    soldOutPerformances: entityMapper.getSoldOutPerformancesInitialValue(
      event.soldOutPerformances
    ),
    description: entityMapper.getRichTextInitialValue(event.description),
    descriptionCredit: event.descriptionCredit || '',
    links: entityMapper.getLinksInitialValue(event.links),
    images: entityMapper.getImagesInitialValue(event.images),
    weSay: event.weSay || '',
    version: event.version || 0
  }
}

// TODO move to the talent/entity module?
function _getTalentsInitialValue (talents) {
  return (talents || []).map(talent => ({
    ...talent,
    key: talent.id,
    roles: (talent.roles || []).join(', '),
    characters: (talent.characters || []).join(', ')
  }))
}

export function mapSubmittedValues (values) {
  const isFree = !eventLib.eventIsPaid(values.costType)
  const costFrom = isFree ? null : parseFloat(values.costFrom)
  const costTo = isFree ? null : parseFloat(values.costTo)

  const bookingNotRequired = !eventLib.bookingRequired(values.bookingType)
  const bookingOpens = bookingNotRequired ? null : values.bookingOpens
  const hasOccurrenceRange = eventLib.occurrenceHasDateRange(
    values.occurrenceType
  )

  return {
    status: values.status,
    name: values.name.trim(),
    eventType: values.eventType,
    occurrenceType: values.occurrenceType,
    dateFrom: hasOccurrenceRange ? values.dateFrom : null,
    dateTo: hasOccurrenceRange ? values.dateTo : null,
    summary: values.summary.trim(),
    costType: values.costType,
    costFrom: costFrom,
    costTo: costTo,
    bookingType: values.bookingType,
    bookingOpens: bookingOpens,
    venueId: values.venue.id,
    venueGuidance: values.venueGuidance.trim(),
    eventSeriesId: values.eventSeries.id || null,
    duration: values.duration.trim(),
    useVenueOpeningTimes: values.useVenueOpeningTimes,
    timesRanges: entityMapper.mapSubmittedTimesRanges(values.timesRanges),
    openingTimes: entityMapper.mapSubmittedOpeningTimes(values.openingTimes),
    additionalOpeningTimes: entityMapper.mapSubmittedAdditionalOpeningTimes(
      values.additionalOpeningTimes
    ),
    specialOpeningTimes: entityMapper.mapSubmittedSpecialOpeningTimes(
      values.specialOpeningTimes
    ),
    openingTimesClosures: entityMapper.mapSubmittedOpeningTimesClosures(
      values.openingTimesClosures
    ),
    performances: entityMapper.mapSubmittedPerformances(values.performances),
    additionalPerformances: entityMapper.mapSubmittedAdditionalPerformances(
      values.additionalPerformances
    ),
    specialPerformances: entityMapper.mapSubmittedSpecialPerformances(
      values.specialPerformances
    ),
    performancesClosures: entityMapper.mapSubmittedPerformancesClosures(
      values.performancesClosures
    ),
    talents: entityMapper.mapSubmittedEventTalents(values.talents),
    mediumTags: values.mediumTags,
    styleTags: values.styleTags,
    audienceTags: values.audienceTags,
    geoTags: values.geoTags,
    weSay: (values.weSay || '').trim(),
    rating: parseInt(values.rating),
    minAge: values.minAge ? parseInt(values.minAge) : null,
    // namedClosures: entityMapper.mapSubmittedNamedClosures(values.namedClosures),
    description: entityMapper.mapSubmittedDescription(values.description),
    descriptionCredit: values.descriptionCredit.trim(),
    links: entityMapper.mapSubmittedLinks(values.links),
    images: entityMapper.mapSubmittedImages(values.images),
    version: values.version + 1
  }
}
