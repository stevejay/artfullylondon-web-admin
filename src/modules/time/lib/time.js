const DAYS = [
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
  'Sunday'
]

export function formatDayNumberForDisplay (dayNumber) {
  return DAYS[parseInt(dayNumber)]
}

export function getTimesRangesOptions (timesRanges) {
  return (timesRanges || []).map(timesRange => {
    return {
      value: timesRange.dateFrom,
      label: 'From ' + timesRange.dateFrom
    }
  })
}
