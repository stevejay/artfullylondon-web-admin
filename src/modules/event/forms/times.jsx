import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Field, reduxForm, formValueSelector } from 'redux-form'

import Form from '_src/shared/components/form'
import FormRow from '_src/shared/components/form/row'
import FormError from '_src/shared/components/form/error'
import FormButtons from '_src/shared/components/form/buttons'
import Divider from '_src/shared/components/divider'
import CheckboxField from '_src/shared/components/checkbox/field'
import {
  TimesField,
  AddTimesRangeForm,
  AddAdditionalOpeningTimeForm,
  AddOpeningTimeForm,
  AddAdditionalPerformanceForm,
  AddPerformanceForm,
  AddSpecialOpeningTimeForm,
  AddSpecialPerformanceForm,
  AddOpeningTimeClosureForm,
  AddPerformanceClosureForm,
  TimesRangeEntry,
  DateAndTimePeriodEntry,
  DayAndTimePeriodEntry,
  DateAndTimeEntry,
  DayAndTimeEntry,
  constraints as timeConstraints,
  actions as timeActions,
  getTimesRangesOptions
} from '_src/modules/time'
import * as timeLib from '_src/shared/lib/time'
import * as eventConstants from '../constants'

export class EditEventTimesForm extends React.Component {
  render () {
    const {
      initialValues,
      useVenueOpeningTimesValue,
      timesRangesValue,
      handleSubmit,
      error,
      submitting,
      isEdit,
      onCancel,
      onPreviousPage
    } = this.props

    const { eventType, occurrenceType, dateFrom, dateTo, venue } = initialValues

    const displayFlags = timeLib.getEventTimesFormDisplayFlags(
      eventType,
      occurrenceType,
      dateFrom,
      dateTo,
      venue,
      useVenueOpeningTimesValue
    )

    // TODO performance issue here:
    const timesRangesOptions = getTimesRangesOptions(timesRangesValue)

    return (
      <Form onSubmit={handleSubmit}>
        {displayFlags.showUseVenueTimesOption &&
          <FormRow>
            <Field
              label='Venue Opening Times'
              name='useVenueOpeningTimes'
              component={CheckboxField}
              checkboxLabel="Use the venue's opening times"
            />
          </FormRow>}
        {displayFlags.showUseVenueTimesOption && <Divider />}
        {displayFlags.showTimesRanges &&
          <FormRow>
            <Field
              label='Times Ranges'
              name='timesRanges'
              component={TimesField}
              formComponent={AddTimesRangeForm}
              itemComponent={TimesRangeEntry}
              constraint={timeConstraints.TIMES_RANGE_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              minDate={dateFrom}
              maxDate={dateTo}
              onSubmit={timeActions.addTimesRange}
              timesRangesOptions={timesRangesOptions}
            />
          </FormRow>}
        {displayFlags.showAdditionalOpeningTimesAsOpeningTimes &&
          <FormRow>
            <Field
              label='Opening Times'
              name='additionalOpeningTimes'
              component={TimesField}
              formComponent={AddAdditionalOpeningTimeForm}
              itemComponent={DateAndTimePeriodEntry}
              constraint={timeConstraints.ADDITIONAL_OPENING_TIME_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              minDate={dateFrom}
              maxDate={dateTo}
              onSubmit={timeActions.addAdditionalOpeningTime}
            />
          </FormRow>}
        {displayFlags.showOpeningTimes &&
          <FormRow>
            <Field
              label='Opening Times'
              name='openingTimes'
              component={TimesField}
              formComponent={AddOpeningTimeForm}
              itemComponent={DayAndTimePeriodEntry}
              constraint={timeConstraints.OPENING_TIME_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              timesRangesOptions={timesRangesOptions}
              onSubmit={timeActions.addOpeningTime}
            />
          </FormRow>}
        {displayFlags.showAdditionalOpeningTimes &&
          <FormRow>
            <Field
              label='Additional Opening Times'
              name='additionalOpeningTimes'
              component={TimesField}
              formComponent={AddAdditionalOpeningTimeForm}
              itemComponent={DateAndTimePeriodEntry}
              constraint={timeConstraints.ADDITIONAL_OPENING_TIME_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              minDate={dateFrom}
              maxDate={dateTo}
              onSubmit={timeActions.addAdditionalOpeningTime}
            />
          </FormRow>}
        {displayFlags.showAdditionalPerformancesAsPerformances &&
          <FormRow>
            <Field
              label='Performances'
              name='additionalPerformances'
              component={TimesField}
              formComponent={AddAdditionalPerformanceForm}
              itemComponent={DateAndTimeEntry}
              constraint={timeConstraints.ADDITIONAL_PERFORMANCE_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              minDate={dateFrom}
              maxDate={dateTo}
              onSubmit={timeActions.addAdditionalPerformance}
            />
          </FormRow>}
        {displayFlags.showPerformances &&
          <FormRow>
            <Field
              label='Performance Times'
              name='performances'
              component={TimesField}
              formComponent={AddPerformanceForm}
              itemComponent={DayAndTimeEntry}
              constraint={timeConstraints.PERFORMANCE_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              timesRangesOptions={timesRangesOptions}
              onSubmit={timeActions.addPerformance}
            />
          </FormRow>}
        {displayFlags.showAdditionalPerformances &&
          <FormRow>
            <Field
              label='Additional Performances'
              name='additionalPerformances'
              component={TimesField}
              formComponent={AddAdditionalPerformanceForm}
              itemComponent={DateAndTimeEntry}
              constraint={timeConstraints.ADDITIONAL_PERFORMANCE_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              minDate={dateFrom}
              maxDate={dateTo}
              onSubmit={timeActions.addAdditionalPerformance}
            />
          </FormRow>}
        {displayFlags.showSpecialOpeningTimes &&
          <FormRow>
            <Field
              label='Special Openings'
              name='specialOpeningTimes'
              component={TimesField}
              formComponent={AddSpecialOpeningTimeForm}
              itemComponent={DateAndTimePeriodEntry}
              constraint={timeConstraints.SPECIAL_OPENING_TIME_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              minDate={dateFrom}
              maxDate={dateTo}
              onSubmit={timeActions.addSpecialOpeningTime}
            />
          </FormRow>}
        {displayFlags.showSpecialPerformances &&
          <FormRow>
            <Field
              label='Special Performances'
              name='specialPerformances'
              component={TimesField}
              formComponent={AddSpecialPerformanceForm}
              itemComponent={DateAndTimeEntry}
              constraint={timeConstraints.SPECIAL_PERFORMANCES_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              minDate={dateFrom}
              maxDate={dateTo}
              onSubmit={timeActions.addSpecialPerformance}
            />
          </FormRow>}
        {displayFlags.showOpeningTimesClosures &&
          <FormRow>
            <Field
              label='Closures'
              name='openingTimesClosures'
              component={TimesField}
              formComponent={AddOpeningTimeClosureForm}
              itemComponent={DateAndTimePeriodEntry}
              constraint={timeConstraints.OPENING_TIME_CLOSURE_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              minDate={dateFrom}
              maxDate={dateTo}
              onSubmit={timeActions.addOpeningTimeClosure}
            />
          </FormRow>}
        {displayFlags.showPerformancesClosures &&
          <FormRow>
            <Field
              label='Closures'
              name='performancesClosures'
              component={TimesField}
              formComponent={AddPerformanceClosureForm}
              itemComponent={DateAndTimeEntry}
              constraint={timeConstraints.PERFORMANCE_CLOSURE_CONSTRAINT}
              parentFormName={eventConstants.EDIT_EVENT_TIMES_FORM_NAME}
              minDate={dateFrom}
              maxDate={dateTo}
              onSubmit={timeActions.addPerformanceClosure}
            />
          </FormRow>}
        <Divider />
        <FormError error={error} />
        <FormButtons
          submitLabel='Next'
          submitting={submitting}
          onCancel={isEdit ? onCancel : null}
          onPrevious={onPreviousPage}
        />
      </Form>
    )
  }
}

EditEventTimesForm.propTypes = {
  initialValues: PropTypes.object.isRequired,
  isEdit: PropTypes.bool.isRequired,
  useVenueOpeningTimesValue: PropTypes.bool,
  timesRangesValue: PropTypes.array,
  submitting: PropTypes.bool.isRequired,
  error: PropTypes.any,
  handleSubmit: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  onPreviousPage: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired
}

const WrappedEditEventTimesForm = reduxForm({
  form: eventConstants.EDIT_EVENT_TIMES_FORM_NAME,
  enableReinitialize: true
})(EditEventTimesForm)

const selector = formValueSelector(eventConstants.EDIT_EVENT_TIMES_FORM_NAME)

export default connect(
  /* istanbul ignore next */
  state => ({
    useVenueOpeningTimesValue: selector(state, 'useVenueOpeningTimes'),
    timesRangesValue: selector(state, 'timesRanges')
  })
)(WrappedEditEventTimesForm)
