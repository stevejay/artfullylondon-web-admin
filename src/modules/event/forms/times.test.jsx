import React from 'react'
import _ from 'lodash'

import { EditEventTimesForm } from './times'
import * as timeLib from '_src/shared/lib/time'
import eventType from '_src/domain/types/event-type'
import occurrenceType from '_src/domain/types/occurrence-type'

it('should render correctly when creating and no display flags are set', () => {
  timeLib.getEventTimesFormDisplayFlags = jest.fn().mockReturnValue({
    showUseVenueTimesOption: false,
    showTimesRanges: false,
    showAdditionalOpeningTimesAsOpeningTimes: false,
    showOpeningTimes: false,
    showAdditionalOpeningTimes: false,
    showAdditionalPerformancesAsPerformances: false,
    showPerformances: false,
    showAdditionalPerformances: false,
    showSpecialOpeningTimes: false,
    showSpecialPerformances: false,
    showOpeningTimesClosures: false,
    showPerformancesClosures: false
  })

  const wrapper = shallow(
    <EditEventTimesForm
      initialValues={{
        eventType: eventType.PERFORMANCE,
        occurrenceType: occurrenceType.BOUNDED,
        dateFrom: '2017/01/01',
        dateTo: '2018/01/01',
        venue: { id: 'some-venue-id' }
      }}
      isEdit={false}
      useVenueOpeningTimesValue={false}
      timesRangesValue={[{ from: '2017/02/02' }]}
      submitting={false}
      imageEditorIsPristine
      handleSubmit={_.noop}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render correctly when editing and all display flags are set', () => {
  timeLib.getEventTimesFormDisplayFlags = jest.fn().mockReturnValue({
    showUseVenueTimesOption: true,
    showTimesRanges: true,
    showAdditionalOpeningTimesAsOpeningTimes: true,
    showOpeningTimes: true,
    showAdditionalOpeningTimes: true,
    showAdditionalPerformancesAsPerformances: true,
    showPerformances: true,
    showAdditionalPerformances: true,
    showSpecialOpeningTimes: true,
    showSpecialPerformances: true,
    showOpeningTimesClosures: true,
    showPerformancesClosures: true
  })

  const wrapper = shallow(
    <EditEventTimesForm
      initialValues={{
        eventType: eventType.PERFORMANCE,
        occurrenceType: occurrenceType.BOUNDED,
        dateFrom: '2017/01/01',
        dateTo: '2018/01/01',
        venue: { id: 'some-venue-id' }
      }}
      isEdit
      useVenueOpeningTimesValue={false}
      timesRangesValue={[]}
      submitting={false}
      imageEditorIsPristine
      handleSubmit={_.noop}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should invoke getEventTimesFormDisplayFlags with the correct values', () => {
  timeLib.getEventTimesFormDisplayFlags = jest.fn().mockReturnValue({})

  shallow(
    <EditEventTimesForm
      initialValues={{
        eventType: eventType.PERFORMANCE,
        occurrenceType: occurrenceType.BOUNDED,
        dateFrom: '2017/01/01',
        dateTo: '2018/01/01',
        venue: { id: 'some-venue-id' }
      }}
      isEdit={false}
      useVenueOpeningTimesValue={false}
      timesRangesValue={[]}
      submitting={false}
      imageEditorIsPristine
      handleSubmit={_.noop}
      onSubmit={_.noop}
      onCancel={_.noop}
      onPreviousPage={_.noop}
      dispatch={_.noop}
    />
  )

  expect(timeLib.getEventTimesFormDisplayFlags).toHaveBeenCalledWith(
    eventType.PERFORMANCE,
    occurrenceType.BOUNDED,
    '2017/01/01',
    '2018/01/01',
    { id: 'some-venue-id' },
    false
  )
})
