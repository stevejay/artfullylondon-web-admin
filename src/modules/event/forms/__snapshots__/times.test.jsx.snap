// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when creating and no display flags are set 1`] = `
<Form
  onSubmit={[Function]}
>
  <Divider
    type="basic"
  />
  <FormError />
  <FormButtons
    onCancel={null}
    onPrevious={[Function]}
    submitLabel="Next"
    submitting={false}
  />
</Form>
`;

exports[`should render correctly when editing and all display flags are set 1`] = `
<Form
  onSubmit={[Function]}
>
  <FormRow>
    <Field
      checkboxLabel="Use the venue's opening times"
      component={[Function]}
      label="Venue Opening Times"
      name="useVenueOpeningTimes"
    />
  </FormRow>
  <Divider
    type="basic"
  />
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "dateFrom": Object {
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "dateTo": Object {
            "dependency": Object {
              "ensure": [Function],
              "message": "Date To is not greater than Date From",
              "test": [Function],
            },
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "label": Object {
            "length": Object {
              "maximum": 20,
              "minimum": 1,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Times Ranges"
      maxDate="2018/01/01"
      minDate="2017/01/01"
      name="timesRanges"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
      timesRangesOptions={Array []}
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "date": Object {
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "from": Object {
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "to": Object {
            "dependency": Object {
              "ensure": [Function],
              "message": "To time is not greater than From time",
            },
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Opening Times"
      maxDate="2018/01/01"
      minDate="2017/01/01"
      name="additionalOpeningTimes"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "day": Object {
            "format": /\\^\\[0123456\\]\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "from": Object {
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "to": Object {
            "dependency": Object {
              "ensure": [Function],
              "message": "To time is not greater than From time",
            },
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Opening Times"
      name="openingTimes"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
      timesRangesOptions={Array []}
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "date": Object {
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "from": Object {
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "to": Object {
            "dependency": Object {
              "ensure": [Function],
              "message": "To time is not greater than From time",
            },
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Additional Opening Times"
      maxDate="2018/01/01"
      minDate="2017/01/01"
      name="additionalOpeningTimes"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "at": Object {
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "date": Object {
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Performances"
      maxDate="2018/01/01"
      minDate="2017/01/01"
      name="additionalPerformances"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "at": Object {
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "day": Object {
            "format": /\\^\\[0123456\\]\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Performance Times"
      name="performances"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
      timesRangesOptions={Array []}
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "at": Object {
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "date": Object {
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Additional Performances"
      maxDate="2018/01/01"
      minDate="2017/01/01"
      name="additionalPerformances"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "audienceTags": Object {
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "date": Object {
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "from": Object {
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "to": Object {
            "dependency": Object {
              "ensure": [Function],
              "message": "To time is not greater than From time",
            },
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Special Openings"
      maxDate="2018/01/01"
      minDate="2017/01/01"
      name="specialOpeningTimes"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "at": Object {
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "audienceTags": Object {
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "date": Object {
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Special Performances"
      maxDate="2018/01/01"
      minDate="2017/01/01"
      name="specialPerformances"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "date": Object {
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
          "from": Object {
            "dependency": Object {
              "ensure": [Function],
              "message": "To time given but From time missing",
              "test": [Function],
            },
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
          },
          "to": Object {
            "dependency": Object {
              "ensure": [Function],
              "message": "To time is not greater than From time",
              "test": [Function],
            },
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Closures"
      maxDate="2018/01/01"
      minDate="2017/01/01"
      name="openingTimesClosures"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      constraint={
        Object {
          "at": Object {
            "format": /\\^\\(\\?:\\[01\\]\\[0-9\\]\\|2\\[0-3\\]\\):\\[0-5\\]\\[0-9\\]\\$\\|\\^\\$/,
          },
          "date": Object {
            "format": /\\^\\[12\\]\\\\d\\\\d\\\\d\\\\/\\[01\\]\\\\d\\\\/\\[0123\\]\\\\d\\$/,
            "presence": Object {
              "disallowEmpty": true,
            },
          },
        }
      }
      formComponent={[Function]}
      itemComponent={[Function]}
      label="Closures"
      maxDate="2018/01/01"
      minDate="2017/01/01"
      name="performancesClosures"
      onSubmit={[Function]}
      parentFormName="EditEventTimes"
    />
  </FormRow>
  <Divider
    type="basic"
  />
  <FormError />
  <FormButtons
    onCancel={[Function]}
    onPrevious={[Function]}
    submitLabel="Next"
    submitting={false}
  />
</Form>
`;
